titus-common/src/main/java/com/netflix/titus/common/network/reverseproxy/grpc/ReverseProxyServerCallHandler.java

```diff
@@ -0,0 +1,139 @@
+        this.channel = channel;
+        this.methodName = methodName;
+    }
+
+    @Override
+    public ServerCall.Listener<Object> startCall(ServerCall<Object, Object> serverCall, Metadata headers) {
+        ClientCall<Object, Object> forwardedCall = channel.newCall(
+                newMethodDescriptorBuilder()
+                        .setType(MethodDescriptor.MethodType.UNARY)
+                        .setFullMethodName(methodName)
+                        .build(),
+                CallOptions.DEFAULT
+        );
+
+        ClientCall.Listener<Object> clientCallListener = new ClientCall.Listener<Object>() {
+            @Override
+            public void onHeaders(Metadata headers) {
+                serverCall.sendHeaders(headers);
+            }
+
+            @Override
+            public void onMessage(Object message) {
+                serverCall.sendMessage(message);
+            }
+
+            @Override
+            public void onClose(Status status, Metadata trailers) {
+                serverCall.close(status, trailers);
+            }
+
+            @Override
+            public void onReady() {
+                serverCall.request(1);
+            }
+        };
+
+        ServerCall.Listener<Object> serverCallListener = new ServerCall.Listener<Object>() {
+
+            private volatile boolean halfClosed = false;
+
+            @Override
+            public void onMessage(Object message) {
+                forwardedCall.request(2);
+
+                try {
+                    forwardedCall.sendMessage(message);
+                } catch (RuntimeException | Error e) {
+                    throw cancelThrow(forwardedCall, e);
+                }
+                onHalfClose();
+            }
+
+            @Override
+            public void onHalfClose() {
+                if (!halfClosed) {
+                    halfClosed = true;
+                    try {
+                        forwardedCall.halfClose();
+                    } catch (RuntimeException | Error e) {
+                        throw cancelThrow(forwardedCall, e);
+                    }
+                }
+            }
+
+            @Override
+            public void onCancel() {
+                try {
+                    forwardedCall.cancel("Client request cancelled", null);
+                } catch (RuntimeException | Error e) {
+                    throw cancelThrow(forwardedCall, e);
+                }
+            }
+
+            @Override
+            public void onComplete() {
+                onHalfClose();
+            }
+
+            @Override
+            public void onReady() {
+                try {
+                    forwardedCall.request(1);
+                } catch (RuntimeException | Error e) {
+                    throw cancelThrow(forwardedCall, e);
+                }
+            }
+        };
+
+        forwardedCall.start(clientCallListener, new Metadata());
+        serverCall.request(1);
```

Comment:
I think unary methods need to do a request(2) to handle the case of multiple responses.

Reply:
Fixed